### PR TITLE
Create mocks structure

### PR DESCRIPTION
**Pull request by BoskiJanusz**
- Modified dependencies.sh to run apt once
- Introduce mock Doctor
- Introduce mock Receptionist
- Introduce factory for Doctor
- Few tests for Doctor
- Few tests for Receptionist

### CODE REVIEW

File: modules/backend/serializer/SerializerHelper.hpp [hunk 1]
```diff
@@ -120,9 +120,10 @@ ::clinic::staff::doctor::Doctor deserializeDoctor(const ::proto_staff::Doctor& p
 ::proto_staff::Staff serializeStaff(const ::clinic::staff::Staff& staff)
 {
     ::proto_staff::Staff proto_staff{};
-    for (const auto& doctor : staff.getDoctors())
+    for (const auto& doctor_ptr : staff.getDoctors())
     {
-        serializeDoctor(doctor, proto_staff.add_doctor());
+        const auto* doctor = dynamic_cast<const clinic::staff::doctor::Doctor*>(doctor_ptr);
+        serializeDoctor(*doctor, proto_staff.add_doctor());
```
Comment (fbtom) [inline modules/backend/serializer/SerializerHelper.hpp:126]: Instead dynamic_casting here, just modify serializeDoctor to take as an argument iDoctor and then reference to structures of passed ptr.

**Review by fbtom — CHANGES_REQUESTED** [review]
Looks good, minor changes requested

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: modules/backend/clinic/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,8 @@
 set(CLINIC_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/Clinic.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/room/Room.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/staff/doctor/Doctor.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/staff/doctors/doctor/Doctor.cpp
```
[the remark this answers is not in the document]
Comment (BoskiJanusz) [reply modules/backend/clinic/CMakeLists.txt:4]: done

File: modules/backend/serializer/SerializerHelper.hpp (hunk 1 above)
Comment (BoskiJanusz) [reply modules/backend/serializer/SerializerHelper.hpp:126]: Done - good idea

File: modules/backend/clinic/staff/Staff.hpp [hunk 3]
```diff
@@ -12,12 +12,20 @@ namespace staff
...
   public:
-    void addDoctor(const doctor::Doctor& doctor);
-    std::vector<doctor::Doctor> getDoctors() const;
+    void addDoctor(doctor::DoctorPtr doctor);
+    doctor::iDoctor* getDoctor(const common::Name& name);
+    std::vector<doctor::iDoctor*> getDoctors() const;
+
+    void addReceptionist(receptionist::ReceptionistPtr receptionist);
+    receptionist::iReceptionist* getReceptionist(const common::Name& name);
+    std::vector<receptionist::iReceptionist*> getReceptionists();
```
[the remark this answers is not in the document]
Comment (BoskiJanusz) [reply modules/backend/clinic/staff/Staff.hpp:21]: done

File: dependencies.sh [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
+
+apt install -y \
+    libgtest-dev libgmock-dev \
+    cmake make g++ \
+    qt6-base-dev \
+    libgl1-mesa-dev \
+    libssl-dev \
+    nlohmann-json3-dev \
+    libgoogle-glog-dev \
+    protobuf-compiler libprotobuf-dev \
```
[the remark this answers is not in the document]
Comment (BoskiJanusz) [reply dependencies.sh:13]: It does

File: modules/backend/serializer/SerializerHelper.hpp [hunk 5]
```diff
@@ -132,8 +132,8 @@ void deserializeStaff(::clinic::staff::Staff& staff, const ::proto_staff::Staff&
 {
     for (const auto& proto_doctor : proto_staff.doctor())
     {
-        auto doctor = deserializeDoctor(proto_doctor);
-        staff.addDoctor(doctor);
+        auto doctor = std::make_unique<clinic::staff::doctor::Doctor>(deserializeDoctor(proto_doctor));
+        staff.addDoctor(std::move(doctor));
```
Comment (fbtom) [inline modules/backend/serializer/SerializerHelper.hpp:136]: I just wonder if we could use doctor's factory here also

**Review by fbtom — APPROVED** [review]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @BoskiJanusz for making course materials better!

**Pull request merged**